### PR TITLE
Make ParseError Much Simpler now that we can use TypeScript

### PR DESCRIPTION
**Pull request by tolmasky**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            |  No |
| Patch: Bug Fix?          | No |
| Major: Breaking Change?  |
| Minor: New Feature?      | No |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | N/A
| Any Dependency Changes?  | No |
| License                  | MIT

As originally promised in the [TypeSafe Errors PR](https://github.com/babel/babel/pull/14320), this PR dramatically simplifies all the `ParseError` machinery now that we are on TypeScript (woooo!!!). From a usability perspective, it is now much easier to define error templates, no longer requiring the use of the weird `_()` flow-workaround-function. So, before you'd have to do something like this:

```typescript
const TSErrors = ParseErrorEnum`typescript` (
  (_: typeof toParseErrorCredentials) => ({
    AbstractMethodHasImplementation: _<{
      methodName: string;
    }>(
      ({ methodName }) =>
        `Method '${methodName}' cannot have an implementation because it is marked abstract.`,
    ),
    AccesorCannotDeclareThisParameter: _(
      "'get' and 'set' accessors cannot declare 'this' parameters.",
    ),
    // ...
  }),
);
```

And now you instead do this:

```typescript
const TSErrors = ParseErrorEnum`typescript` (
  AbstractMethodHasImplementation: ({ methodName }:  { methodName: string }) =>
    `Method '${methodName}' cannot have an implementation because it is marked abstract.`,
  AccesorCannotDeclareThisParameter: "'get' and 'set' accessors cannot declare 'this' parameters.",
  // ...
});
```

In other words, we've kept the simplicity of the old object reason-code to message mapping, but with real types. This was of course the original goal, but many flow bugs prevented us from doing this as cleanly, so we needed these weird "carrier" functions (like `toParseErrorCredentials`) to shuttle the types around. TypeScript is much better at tracking the types everywhere and so we no longer need them and can so we can just use normal objects.

Aside from that, there are just a few random fixes to get rid of `ts-expect-error`s, etc., and updating comments to reflect the new code.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14796"></a>

### CODE REVIEW

File: packages/babel-parser/src/parse-error/credentials.ts [hunk 1]
```diff
@@ -1,10 +1,7 @@
-export const ParseErrorCodes = Object.freeze({
-  SyntaxError: "BABEL_PARSER_SYNTAX_ERROR",
-  SourceTypeModuleError: "BABEL_PARSER_SOURCETYPE_MODULE_REQUIRED",
-});
-
-export type ParseErrorCode =
-  typeof ParseErrorCodes[keyof typeof ParseErrorCodes];
+export enum ParseErrorCode {
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parse-error/credentials.ts:1]: Can you use `const enum` instead? We don't use the reverse mapping anyway.

**Review by nicolo-ribaudo — COMMENTED** [review]
Awesome, thank you!

File: packages/babel-parser/src/parse-error/credentials.ts (hunk 1 above)
Comment (tolmasky) [reply packages/babel-parser/src/parse-error/credentials.ts]: Good catch! Done: https://github.com/babel/babel/pull/14796/commits/c5f4bba84657ec7dbcc8cea7a37ff77cf2be74cc

**Review by liuxingbaoyu — APPROVED** [review]
This PR is great and it looks so much better now!

**Review by JLHwung — APPROVED** [review]
This PR looks great to me.
